# Rod Pump Group #2
# Exploratory Data Analysis

In [1]:
import pandas as pd
import seaborn as sns
%matplotlib inline

raw_data = pd.read_csv('../rodpump_raw.csv')

In [2]:
for column in list(raw_data):
    print(column)

roduid
UWI
NODEID
IDWELL
tbguid
lifetime_start
lifetime_end
IDRECJOBPULL
REPORTTO
FAILSTART
FAILURETYPE
H2S_CONCENTRATION
PrimarySetpoint
SecondarySetpoint
StrokeLength
GrossStrokeLength
Fillage
YesterdaysAverageSPM
bha_configuration
chemgroup1_any
chemgroup1_all
chemgroup2_any
chemgroup2_all
chemgroup3_any
chemgroup3_all
max_unguided_dls
dls_high_in_hole
gas_anchor_length
MAX_INCLINATION
wellbore_category
manual_scale
packer_vs_tac
AVG_PRESS_FLOWLINE
AVG_PRESSURE_TUBING
AVG_PRESSURE_CASING
AVG_DIFFERENTIAL_PRESSURE
AVG_OIL_VOLUME
AVG_WATER_VOLUME
AVG_LIQUID_VOLUME
AVG_WATERSG
rod_sinker_type
rod_has_guides
rod_make
rod_apigrade
ROUTE
overall_max_sideload
shallow_max_sideload
max_unguided_sideload
DESANDDEGAS_TYP
CHROME_LENGTH
ENDURALLOY_LENGTH
POLY_LENGTH
NIPPLE_SET_DEPTH
pump_bore
gasanchor_od


## List rows with repeated values

In [3]:
raw_rows = raw_data.shape[0]
for (columnName, columnData) in raw_data.iteritems():
    mask = raw_data[columnName].notnull()
    if (columnData[mask].unique().size < raw_rows):
        print(columnName + "," + str(columnData.unique().size))

UWI,633
NODEID,632
IDWELL,633
tbguid,1815
lifetime_start,1792
lifetime_end,1467
IDRECJOBPULL,1693
REPORTTO,5
FAILSTART,1273
FAILURETYPE,4
H2S_CONCENTRATION,51
PrimarySetpoint,43
SecondarySetpoint,18
StrokeLength,303
GrossStrokeLength,1485
Fillage,1391
YesterdaysAverageSPM,238
bha_configuration,6
chemgroup1_any,2
chemgroup1_all,2
chemgroup2_any,2
chemgroup2_all,2
chemgroup3_any,2
chemgroup3_all,2
max_unguided_dls,652
dls_high_in_hole,533
gas_anchor_length,282
MAX_INCLINATION,390
wellbore_category,5
manual_scale,2
packer_vs_tac,6
AVG_PRESS_FLOWLINE,1659
AVG_PRESSURE_TUBING,2425
AVG_PRESSURE_CASING,2475
AVG_DIFFERENTIAL_PRESSURE,2429
AVG_OIL_VOLUME,2554
AVG_WATER_VOLUME,2549
AVG_LIQUID_VOLUME,2558
AVG_WATERSG,78
rod_sinker_type,4
rod_has_guides,2
rod_make,7
rod_apigrade,7
ROUTE,20
overall_max_sideload,1680
shallow_max_sideload,1675
max_unguided_sideload,1662
DESANDDEGAS_TYP,6
CHROME_LENGTH,20
ENDURALLOY_LENGTH,370
POLY_LENGTH,31
NIPPLE_SET_DEPTH,1702
pump_bore,5
gasanchor_od,5


## Identify potential categorical columns
Potential categorical columns will be defined as any column with a small set of (n<10) frequently repeated values

In [4]:
# List columns with a small set of frequently repeated values, and their frequencies
raw_rows = raw_data.shape[0]
for (columnName, columnData) in raw_data.iteritems():
    if (columnData.unique().size <= 10):
        print(columnName)
        print(columnData.value_counts(ascending=False, dropna=False))
        print('')

REPORTTO
Tubing             899
NaN                871
Sucker Rod Pump    612
Rods               210
Liner (Casing)       4
Name: REPORTTO, dtype: int64

FAILURETYPE
Tubing             899
NaN                875
Sucker Rod Pump    612
Rods               210
Name: FAILURETYPE, dtype: int64

bha_configuration
TAC_ABOVE_NIP        914
TAC_BELOW_NIP        777
PACKER_DONNAN        626
PACKER_TAC_DONNAN    193
NaN                   46
NOSEP_PACKER          40
Name: bha_configuration, dtype: int64

chemgroup1_any
0    1375
1    1221
Name: chemgroup1_any, dtype: int64

chemgroup1_all
0    2378
1     218
Name: chemgroup1_all, dtype: int64

chemgroup2_any
0    1990
1     606
Name: chemgroup2_any, dtype: int64

chemgroup2_all
0    2372
1     224
Name: chemgroup2_all, dtype: int64

chemgroup3_any
0    2422
1     174
Name: chemgroup3_any, dtype: int64

chemgroup3_all
0    2514
1      82
Name: chemgroup3_all, dtype: int64

wellbore_category
Vertical       1647
LowTangent      707
HighTangent     18

#### Columns 'UWI', 'NODEID', and 'IDWELL' appear to be redundant
Each column contains the same number of unique values, and the frequencies of the top five items are the same

In [5]:
raw_data['UWI'].value_counts(ascending=False, dropna=False)

540-51-3641    15
352-78-9074    11
859-66-6410    10
427-18-9490    10
292-26-5777    10
               ..
255-56-7105     1
790-49-1299     1
286-03-7584     1
021-28-3780     1
512-20-1013     1
Name: UWI, Length: 633, dtype: int64

In [6]:
raw_data['IDWELL'].value_counts(ascending=False, dropna=False)

FOKS36813367797224    15
XTDY59752018206495    11
GKWW90620445815318    10
EDOR20136896483790    10
KGTI45658951785364    10
                      ..
SVJE58183882752026     1
FYVN74590548879605     1
CPTT74954618126083     1
YOWC31575117085366     1
OINX89179940885938     1
Name: IDWELL, Length: 633, dtype: int64

In [7]:
raw_data['NODEID'].value_counts(ascending=False, dropna=False)

catch-gas-bill-cost    15
prepare-response       11
senior-help-threat     10
stock-according        10
ask-until-light        10
                       ..
person-same             1
heavy-cold-they         1
pass-everything         1
much-hair-explain       1
nothing-beyond-less     1
Name: NODEID, Length: 632, dtype: int64

#### There is no clear structure or meaning to 'lifetime_start' and 'lifetime_end'
Only 1080 out of 2596 rows have a non-zero 'lifetime_end'

In [8]:
mask = (raw_data['lifetime_end'] == '00:00.0')
raw_data.shape[0] - raw_data['lifetime_end'][mask].count()

2596

There are only 52 unique values for 'lifetime_end', and zero occurs most frequently in 1516 out of 2596 cases.  37:11.3 is another extremely frequent occurance.
#### 37:11.3 corresponds to April 15, 2020.

In [9]:
print(raw_data['lifetime_end'].unique().size)
raw_data['lifetime_end'].value_counts(ascending=False, dropna=False).head(10)

1467


4/15/20    593
9/14/17      5
6/1/19       4
7/16/19      4
9/11/17      4
5/10/18      4
4/2/19       4
2/11/19      4
3/5/20       4
8/1/17       4
Name: lifetime_end, dtype: int64

In 298 cases, 'lifetime_end' is a smaller value than 'lifetime_start'

In [10]:
raw_data['lifetime_start'][raw_data['lifetime_start'] > raw_data['lifetime_end']].count()

1270

#### There is some correlation between 'FAILURETYPE', 'REPORTTO' and 'lifetime_end'
'FAILURETYPE' and 'REPORTTO' appear to be essentially redundant.  We will use 'REPORTTO' in our analysis because it includes the additional 'Linear (Casting)' field.

In [11]:
raw_data['FAILURETYPE'].value_counts(ascending=False, dropna=False)

Tubing             899
NaN                875
Sucker Rod Pump    612
Rods               210
Name: FAILURETYPE, dtype: int64

In [12]:
raw_data['REPORTTO'].value_counts(ascending=False, dropna=False)

Tubing             899
NaN                871
Sucker Rod Pump    612
Rods               210
Liner (Casing)       4
Name: REPORTTO, dtype: int64

In 574 out of 591 occurances of '37:11.3' in 'lifetime_end', 'FAILURETYPE' is null.

In [13]:
mask = (raw_data['lifetime_end'] == '37:11.3') & (raw_data['FAILURETYPE'].isnull())
raw_data['lifetime_end'][mask].size

0

In 38 out of 263 occurances of '30:00.0' in 'lifetime_end', 'FAILURETYPE' is null.

In [14]:
mask = (raw_data['lifetime_end'] == '30:00.0') & (raw_data['FAILURETYPE'].isnull())
raw_data['lifetime_end'][mask].size

0

#### There is also correlation between 'IDRECJOBPULL' and 'FAILURETYPE'
For all 871 occurances of 'IDRECJOBPULL' = 'WJDB83445325478746', 'FAILURETYPE' is null.

In [15]:
mask = (raw_data['IDRECJOBPULL'] == 'WJDB83445325478746') & (raw_data['FAILURETYPE'].isnull())
raw_data['lifetime_end'][mask].size

871

In [16]:
mask = (raw_data['IDRECJOBPULL'] == 'WJDB83445325478746') & ~(raw_data['FAILURETYPE'].isnull())
raw_data['lifetime_end'][mask].size

0

#### Values of 'FAILSTART' also correlate with 'IDRECJOBPULL' and 'FAILURETYPE'
Rows with NaN values in 'FAILSTART' have 'IDRECJOBPULL' = 'WJDB83445325478746' and 'FAILURETYPE' = NaN.

In [17]:
raw_data['FAILSTART'].value_counts(ascending=False, dropna=False)

NaN         871
3/1/13        6
12/6/18       5
5/1/17        5
8/1/12        5
           ... 
8/12/14       1
1/9/07        1
4/23/18       1
11/24/14      1
1/21/15       1
Name: FAILSTART, Length: 1273, dtype: int64

In [18]:
mask = (raw_data['IDRECJOBPULL'] == 'WJDB83445325478746') & (raw_data['FAILSTART'].isnull())
raw_data['lifetime_end'][mask].size

871

### Analyze potential quantitative variables

In [19]:
for targetColumn in {'H2S_CONCENTRATION','PrimarySetpoint','SecondarySetpoint','StrokeLength','GrossStrokeLength','Fillage','YesterdaysAverageSPM','max_unguided_dls','dls_high_in_hole','gas_anchor_length','MAX_INCLINATION','AVG_PRESS_FLOWLINE','AVG_PRESSURE_TUBING','AVG_PRESSURE_CASING','AVG_DIFFERENTIAL_PRESSURE','AVG_OIL_VOLUME','AVG_WATER_VOLUME','AVG_LIQUID_VOLUME','AVG_WATERSG','ROUTE','overall_max_sideload','shallow_max_sideload','max_unguided_sideload','CHROME_LENGTH','ENDURALLOY_LENGTH','POLY_LENGTH','NIPPLE_SET_DEPTH'}:
    print("Column name: " + targetColumn)
    print("(Min,Mean,Max) = (" + str(raw_data[targetColumn].min()) + "," + str(raw_data[targetColumn].mean()) + "," + str(raw_data[targetColumn].max()) + ")")
    count_unique = raw_data[targetColumn].unique().size
    print("Count unique = " + str(count_unique))
    if (count_unique < 50):
          print(raw_data[targetColumn].value_counts(ascending=False, dropna=False).head(15))
    print("")

Column name: H2S_CONCENTRATION
(Min,Mean,Max) = (0.0,10.201564714946072,16500.0)
Count unique = 51

Column name: AVG_PRESSURE_CASING
(Min,Mean,Max) = (0.0,115.77554690133648,2322.5)
Count unique = 2475

Column name: AVG_PRESS_FLOWLINE
(Min,Mean,Max) = (0.337778,57.97728262793498,247.787692)
Count unique = 1659

Column name: AVG_WATERSG
(Min,Mean,Max) = (1.04,1.0725440943905071,1.2)
Count unique = 78

Column name: max_unguided_dls
(Min,Mean,Max) = (0.0,2.7283300922964275,92.34)
Count unique = 652

Column name: MAX_INCLINATION
(Min,Mean,Max) = (0.0,4.690494699646643,53.4)
Count unique = 390

Column name: SecondarySetpoint
(Min,Mean,Max) = (0.0,57.261142498459165,100.0)
Count unique = 18
60.000000    1700
65.000000     606
0.000000      173
50.000000      32
70.000000      23
61.000000      11
66.000000      10
62.000000      10
55.000000       7
64.000000       5
75.000000       5
63.000000       5
90.000000       3
59.000000       2
40.925926       1
Name: SecondarySetpoint, dtype: int6

#### There is some correlation between 'MAX_INCLINATION' and 'GrossStrokeLenth'

In [20]:
raw_data[['MAX_INCLINATION','GrossStrokeLength']].corr()

MAX_INCLINATION  GrossStrokeLength
MAX_INCLINATION           1.000000           0.304075
GrossStrokeLength         0.304075           1.000000

#### There is significant correlation within the 'sideload' derivatives and with 'rod_has_guides'

In [21]:
raw_data[['overall_max_sideload','shallow_max_sideload','max_unguided_sideload','rod_has_guides']].corr()

overall_max_sideload  shallow_max_sideload  \
overall_max_sideload               1.000000              0.970676   
shallow_max_sideload               0.970676              1.000000   
max_unguided_sideload              0.833359              0.811716   
rod_has_guides                     0.344645              0.328094   

                       max_unguided_sideload  rod_has_guides  
overall_max_sideload                0.833359        0.344645  
shallow_max_sideload                0.811716        0.328094  
max_unguided_sideload               1.000000        0.059613  
rod_has_guides                      0.059613        1.000000

There are five cases where the max unguided sideload is greater than either overall max sideload or shallow max sideload, and the rod does not have guides.  Three out of these five have failed.

In [22]:
mask = ((raw_data['overall_max_sideload'] > raw_data['max_unguided_sideload']) | (raw_data['shallow_max_sideload'] > raw_data['max_unguided_sideload'])) & (raw_data['rod_has_guides'] == False)
raw_data[mask]

roduid          UWI               NODEID  \
287   GB24TAFQ54821960591047  121-78-4988     room-last-within   
542   GB35AQUB80942352555022  207-01-3041  ready-rather-camera   
1390  GB81GLWA06155617601586  480-46-7415    can-be-part-money   
1518  GB86RWVS33933117555905  525-71-8998   there-place-school   
2053  GB46CIWB55366663844735  703-17-3023            drug-yeah   

                  IDWELL                  tbguid lifetime_start lifetime_end  \
287   LDJE37247757884478  GB87KVVK33849452527539        1/16/15      9/23/15   
542   YHRJ80622738388627  GB94GJFF78939763421069         5/4/16       8/9/16   
1390  JUYA22128643062093  GB92YHND30694442051946        3/12/15      7/14/15   
1518  AUVF72665292144382  GB77YGZC10630744551715        11/9/19      4/15/20   
2053  MOWF37867415064285  GB83ZFCY47507370910409       11/14/17       3/5/20   

            IDRECJOBPULL REPORTTO FAILSTART  ... overall_max_sideload  \
287   MYXJ15086636721414   Tubing   9/13/15  ...               345.33   
542   WJDB83445325478746      NaN       NaN  ...               124.19   
1390  WZPX94370985238597   Tubing   6/25/15  ...               229.09   
1518  WJDB83445325478746      NaN       NaN  ...               109.08   
2053  UDVB65148573611157     Rods    1/8/20  ...               212.18   

      shallow_max_sideload  max_unguided_sideload  DESANDDEGAS_TYP  \
287                 345.33                   7.70       Miller LLC   
542                 112.29                   3.09       Miller LLC   
1390                229.09                  12.33          UNKNOWN   
1518                109.08                   0.00       Miller LLC   
2053                212.18                 146.18  West-Washington   

      CHROME_LENGTH  ENDURALLOY_LENGTH  POLY_LENGTH  NIPPLE_SET_DEPTH  \
287             0.0               0.00          0.0            9794.4   
542             0.0               0.00          0.0            9502.2   
1390            0.0               0.00          0.0            9997.4   
1518            0.0               0.00          0.0            9783.6   
2053            0.0             124.59          0.0            9610.4   

     pump_bore  gasanchor_od  
287       1.75           NaN  
542       1.75           NaN  
1390       1.5           NaN  
1518       1.5           4.5  
2053      1.75           NaN  

[5 rows x 55 columns]

### Identify other potentially significant variables
We believe 'rod_make' contains alias' for rod pump manufacturers

In [23]:
print(str(raw_data['rod_make'].unique().size) + ' unique values')
raw_data['rod_make'].value_counts(ascending=False, dropna=False)

7 unique values


mesh enterprise portals                 1433
strategize next-generation users        1007
synthesize holistic partnerships         128
incentivize vertical systems              10
visualize leading-edge initiatives        10
extend customized interfaces               5
iterate bricks-and-clicks e-business       3
Name: rod_make, dtype: int64

In [24]:
print(str(raw_data['rod_apigrade'].unique().size) + ' unique values')
raw_data['rod_apigrade'].value_counts(ascending=False, dropna=False)

7 unique values


SpecialService    1695
D                  678
Unknown            129
K                   79
Fiberglass           9
NaN                  5
C                    1
Name: rod_apigrade, dtype: int64

#### Not all manufacturers produce rod pumps of every api_grade
Strategize, mesh, and synthesize are the biggest players in the market

In [25]:
for manufacturer in raw_data['rod_make'].unique():
    print("Manufacturer: " + str(manufacturer))
    mask = (raw_data['rod_make'] == manufacturer)
    print(raw_data['rod_apigrade'][mask].value_counts(ascending=False, dropna=False))
    print('')

Manufacturer: mesh enterprise portals
SpecialService    1254
D                  175
Fiberglass           3
C                    1
Name: rod_apigrade, dtype: int64

Manufacturer: synthesize holistic partnerships
Unknown    128
Name: rod_apigrade, dtype: int64

Manufacturer: strategize next-generation users
D                 499
SpecialService    431
K                  77
Name: rod_apigrade, dtype: int64

Manufacturer: visualize leading-edge initiatives
Fiberglass        6
D                 2
SpecialService    1
Unknown           1
Name: rod_apigrade, dtype: int64

Manufacturer: extend customized interfaces
NaN    5
Name: rod_apigrade, dtype: int64

Manufacturer: incentivize vertical systems
SpecialService    9
D                 1
Name: rod_apigrade, dtype: int64

Manufacturer: iterate bricks-and-clicks e-business
K    2
D    1
Name: rod_apigrade, dtype: int64



In [26]:
print(str(raw_data['bha_configuration'].unique().size) + ' unique values')
raw_data['bha_configuration'].value_counts(ascending=False, dropna=False)

6 unique values


TAC_ABOVE_NIP        914
TAC_BELOW_NIP        777
PACKER_DONNAN        626
PACKER_TAC_DONNAN    193
NaN                   46
NOSEP_PACKER          40
Name: bha_configuration, dtype: int64

In [27]:
print(str(raw_data['GrossStrokeLength'].unique().size) + ' unique values')
raw_data['GrossStrokeLength'].value_counts(ascending=False, dropna=False)

1485 unique values


NaN           1096
0.000000         7
152.000000       4
160.000000       2
148.000000       2
              ... 
136.882648       1
129.586461       1
118.111931       1
130.378703       1
140.199398       1
Name: GrossStrokeLength, Length: 1485, dtype: int64

In [28]:
raw_data.shape

(2596, 55)